### PR TITLE
Improve example rules

### PR DESCRIPTION
**Pull request by tsteenbe**
Please see individual commits for details.

Note: This is only part 1 of improving the rules, will submit part 2 when this PR is merged.

### CODE REVIEW

File: evaluator-rules/src/main/resources/example.rules.kts [hunk 1]
```diff
@@ -53,36 +53,873 @@ val handledLicenses = listOf(
...
+    }
+
 enum class PolicyRules() {
     OSS_PROJECT,
     PROPRIETARY_PROJECT
 }
 
+fun getArtifactOrUrlName(url: String): String {
+    if (url.isNullOrBlank()) {
+         return "name of the artifact"
```
Comment (fviernau) [inline evaluator-rules/src/main/resources/example.rules.kts:109]: could be done in one line like:
```
"name of the artifact".takeIf { url.isNullOrBlank()url.isNullOrBlank() } ?: "${url.substringAfterLast("/")}"
```

**Review by fviernau — CHANGES_REQUESTED** [review]
I haven't reviewed all yet as it is quite large. I'll continue later on...

File: evaluator-rules/src/main/resources/example.rules.kts [hunk 2]
```diff
@@ -54,17 +63,914 @@ val handledLicenses = listOf(
...
+            return """
+                |For this violation, there is no recommended solution.
+                |""".trimMargin()
+        }
+    } else {
+        // Violation is thrown for one of the project's dependencies.
+        if (licenseSource == LicenseSource.DETECTED) {
+            // Violation thrown for license detected by the scanner in the source code of the dependency.
+            return """
+                |${resolveViolationInDependencySourceCodeText(pkg, license)}
```
Comment (sschuberth) [inline evaluator-rules/src/main/resources/example.rules.kts:286]: Just double-checking: It is correct to the the margin character `|` here, although the text returned by `resolveViolationInDependency*` already contains the margin character itself?

Comment (tsteenbe) [reply evaluator-rules/src/main/resources/example.rules.kts]: Fixed it to be a direct return, was left-over after I moved the text that was there into `resolveViolation` functions.

File: evaluator-rules/src/main/resources/example.rules.kts [hunk 3]
```diff
@@ -53,6 +53,27 @@ val handledLicenses = listOf(
...
+enum class PolicyRules() {
+    OSS_PROJECT,
+    PROPRIETARY_PROJECT
+}
+
+/**
+ * Function to enable policy rules based on label passed to ORT.
+ */
+fun getEnabledPolicyRules(): PolicyRules =
+    if (isLabeled("project", "oss-project")) {
```
[the remark this answers is not in the document]
Comment (fviernau) [reply evaluator-rules/src/main/resources/example.rules.kts]: it's not yet fixed. Forgot to push again?

File: evaluator-rules/src/main/resources/example.rules.kts [hunk 4]
```diff
@@ -53,36 +53,855 @@ val handledLicenses = listOf(
...
- * Function to return Markdown-formatted text to aid users with resolving violations.
+ * Function returning true or false if for given package id source artifact was scanned.
+ */
+fun isSourceArtifactScanned(id: Identifier): Boolean =
+    ortResult.getScanResultsForId(id).any { it.provenance is ArtifactProvenance }
+
+/**
+ * Function returning true or false if for given package id code repository was scanned.
  */
+fun isVcsScanned(id: Identifier): Boolean =
```
Comment (fviernau) [inline evaluator-rules/src/main/resources/example.rules.kts:233]: Return true if and only if the [ortResult] contains a scan result for the VCS of the package denoted by [id].

File: evaluator-rules/src/main/resources/example.rules.kts [hunk 5]
```diff
@@ -53,36 +53,855 @@ val handledLicenses = listOf(
...
+    ortResult.getScanResultsForId(id).any { it.provenance is RepositoryProvenance }
 
+/**
+ * Function returning coordinates string without version for given package id. 
+ *
+ * For example, this function will return 'PyPI:flower' for package id 'PyPI::flower:0.9.7'
+ * and 'Maven:org.antlr:antlr4' for 'Maven:org.antlr:antlr4:4.0.0'.
+ *
+ */
+fun Identifier.toCoordinatesWithoutVersion() = "$type:$namespace:$name"
```
Comment (fviernau) [inline evaluator-rules/src/main/resources/example.rules.kts:243]: Return the coordinates without the version.

File: evaluator-rules/src/main/resources/example.rules.kts [hunk 6]
```diff
@@ -53,36 +53,855 @@ val handledLicenses = listOf(
...
+    license: String,
+    licenseSource: LicenseSource,
+    @Suppress("UNUSED_PARAMETER") severity: Severity
+): String {
+    if (ortResult.isProject(pkg.id)) {
+        // Violation is flagged for the project scanned.
+        if (licenseSource == LicenseSource.DETECTED) {
+            // License is detected by the scanner in the source code of the project.
+            return "${resolveViolationInSourceCodeText(pkg, license)}"
+        } else {
```
Comment (fviernau) [inline evaluator-rules/src/main/resources/example.rules.kts:263]: Please drop the `else` in line 240, 244 and 249 (but keep the code residing in the else blocks.

File: evaluator-rules/src/main/resources/example.rules.kts [hunk 7]
```diff
@@ -53,36 +53,855 @@ val handledLicenses = listOf(
...
+                |
+                $createIssueText
+                |""".trimMargin()
+        }
+    }
+}
+
+fun PackageRule.howToFixUnmappedDeclaredLicense(
+    license: String,
+    @Suppress("UNUSED_PARAMETER") severity: Severity
```
Comment (fviernau) [inline evaluator-rules/src/main/resources/example.rules.kts:341]: Please drop this unused parameter.

File: evaluator-rules/src/main/resources/example.rules.kts [hunk 8]
```diff
@@ -53,36 +53,855 @@ val handledLicenses = listOf(
...
+                |
+                |   ```
+                |   curations: Conclude license for ${pkg.id.toCoordinates()}
+                |
+                |   $relatesToIssueText
+                |   ```
+                |
+                |   - Once your $ortCurationsYmlFileConcludedLicenseMdLink is merged, re-scan to see if the violation has been resolved.
+                |"""
+        } else {
```
Comment (fviernau) [inline evaluator-rules/src/main/resources/example.rules.kts:495]: Please drop the else to reduce level of nesting.

Same in several functions below.

File: evaluator-rules/src/main/resources/example.rules.kts (hunk 6 above)
Comment (tsteenbe) [reply evaluator-rules/src/main/resources/example.rules.kts:264]: I actually deliberately did to make it easier for non-technical people such a maybe legal counsels understand the policy.

File: evaluator-rules/src/main/resources/example.rules.kts [hunk 9]
```diff
@@ -262,5 +259,67 @@ val ruleSet = ruleSet(ortResult, licenseInfoResolver, resolutionProvider) {
...
+}
+
+fun RuleSet.ossProjectRules() {
+}
+
+val ruleSet = ruleSet(ortResult, licenseInfoResolver, resolutionProvider) {
+    commonRules()
+    when (getEnabledPolicyRules()) {
+        PolicyRules.PROPRIETARY_PROJECT -> proprietaryProjectRules()
+        PolicyRules.OSS_PROJECT -> ossProjectRules()
```
[the remark this answers is not in the document]
Comment (tsteenbe) [reply evaluator-rules/src/main/resources/example.rules.kts:1240]: One of the selling points of ORT is that one can use it for compliance and contribution processes which is why I included OSS_PROJECT rules. Still a bit under developed but in my follow up PRs I will add more meat.

File: evaluator-rules/src/main/resources/example.rules.kts [hunk 10]
```diff
@@ -53,36 +53,855 @@ val handledLicenses = listOf(
...
+val ortCurationsYmlVcsPathMdLink = "[VCS path curation](https://github.com/oss-review-toolkit/ort/blob/main/docs/config-file-curations-yml.md)"
+val ortCurationsYmlVcsUrlMdLink = "[VCS URL curation](https://github.com/oss-review-toolkit/ort/blob/main/docs/config-file-curations-yml.md)"
+var ortLicenseFindingCurationReasonMdLink = "[LicenseFindingCurationReason.kt](https://github.com/oss-review-toolkit/ort/blob/main/model/src/main/kotlin/config/LicenseFindingCurationReason.kt)"
+val ortPackageConfigurationFileMdLink = "[package configuration](https://github.com/oss-review-toolkit/ort/blob/main/docs/config-file-package-configuration-yml.md)"
+var ortPathExcludeReasonMdLink = "[PathExcludeReason.kt](https://github.com/oss-review-toolkit/ort/blob/main/model/src/main/kotlin/config/PathExcludeReason.kt)"
+val ortResolutionsYmlRuleViolationMdLink = "[rule violation resolution](https://github.com/oss-review-toolkit/ort/blob/main/docs/config-file-resolutions-yml.md#resolving-policy-rule-violations)"
+val ortScanIssueTrackerName = "FOSS JIRA"
```
[the remark this answers is not in the document]
Comment (tsteenbe) [reply evaluator-rules/src/main/resources/example.rules.kts]: Lots of potential ORT users use JIRA, included this to help them better understand how they could build a ORT + JIRA process.

File: evaluator-rules/src/main/resources/example.rules.kts (hunk 8 above)
Comment (tsteenbe) [reply evaluator-rules/src/main/resources/example.rules.kts]: Leaving it as in my follow up PR I will build on top of this.

File: evaluator-rules/src/main/resources/example.rules.kts [hunk 11]
```diff
@@ -1219,7 +1212,7 @@ fun RuleSet.unmappedDeclaredLicenseRule() {
 }
 
 fun RuleSet.commonRules() {
-    unhandledLicenseRule()
+    // unhandledLicenseRule()
```
Comment (fviernau) [inline evaluator-rules/src/main/resources/example.rules.kts:1215]: Left over from debugging?

Comment (tsteenbe) [reply evaluator-rules/src/main/resources/example.rules.kts]: Yes, fixed it.

**Review by fviernau — APPROVED** [review]

**Pull request merged**